File: core/js/src/xact-ids.ts
import { TransactionId } from "./db_fields";

const TOP_BITS = BigInt("0x0DE1") << BigInt(48);
const MOD = BigInt(1) << BigInt(64);
const COPRIME = BigInt("205891132094649");
const COPRIME_INVERSE = BigInt("1522336535492693385");

function modularMultiply(value: bigint, prime: bigint) {
  return (value * prime) % MOD;
}

export function prettifyXact(valueString: TransactionId): string {
  const value = BigInt(valueString);
  const encoded = modularMultiply(value, COPRIME);
  return encoded.toString(16).padStart(16, "0");
}

export function loadPrettyXact(encodedHex: string): TransactionId {
  if (encodedHex.length !== 16) {
    return encodedHex;
  }
  const value = BigInt(`0x${encodedHex}`);
  const multipliedInverse = modularMultiply(value, COPRIME_INVERSE);
  const withTopBits = TOP_BITS | multipliedInverse;
  return withTopBits.toString();
}

File: core/py/src/braintrust_core/xact_ids.py
TOP_BITS = 0x0DE1 << 48

MOD = 2**64
COPRIME = 205891132094649
COPRIME_INVERSE = 1522336535492693385


def modular_multiply(value: int, prime: int):
    return (value * prime) % MOD


def prettify_xact(value: int | str) -> str:
    encoded = modular_multiply(int(value), COPRIME)
    return hex(encoded)[2:].rjust(16, "0")


def load_pretty_xact(encoded_hex: str) -> str:
    value = int(encoded_hex, 16)
    multiplied_inverse = modular_multiply(value, COPRIME_INVERSE)
    with_top_bits = TOP_BITS | multiplied_inverse
    return str(with_top_bits)
